src/python/pants/pantsd/pants_daemon.py

```diff
@@ -131,15 +133,32 @@ def __init__(
 
         self._logger = logging.getLogger(__name__)
 
-    @staticmethod
-    def _close_stdio():
-        """Close stdio streams to avoid output in the tty that launched pantsd."""
-        for fd in (sys.stdin, sys.stdout, sys.stderr):
+    def _close_stdio(self, log_path: Path):
+        """Close stdio and append to a log path instead.
+
+        The vast majority of Python-level IO will be re-routed to thread-local destinations by
+        `initialize_stdio`, but we close stdio to avoid any stray output in the tty that launched
+        pantsd.
+
+        Rather than leaving 0, 1, 2 dangling though, we open replacements as a backstop for fatal
+        errors or unmodified code (such as Rust panic handlers) that might expect them to be valid
+        file handles.
+        """
+        for attr, writable in (("stdin", False), ("stdout", True), ("stderr", True)):
+            # Close the old.
+            fd = getattr(sys, attr)
             file_no = fd.fileno()
             fd.flush()
             fd.close()
             os.close(file_no)
```

Comment:
Instead of `os.close(file_no)` then `open` and relying on non-threadedness to get you back the same `file_no`, how about dropping the `os.close` and using the more explicit and safe `os.dup2(new_fd, fileno)`?